File: Client/LegacyWebViewBackForward.m
#import "LegacyWebViewBackForward.h"

@interface LegacyBackForwardListItem()
@end


@implementation LegacyBackForwardListItem 

- (NSURL*)url
{
  return self.writableUrl;
}

- (NSURL*)initialUrl
{
  return self.writableInitialUrl;
}

- (NSString*)title
{
  return self.writableTitle;
}
//
//- (void)writableSetUrl:(NSURL*)url
//{
//  self.writableUrl = url;
//}
//
//- (void)writableInitialUrl:(NSURL*)initial
//{
//  self.writableInitialUrl = initial;
//}
//
//- (void)writableTitle:(NSString*)title
//{
//  self.writableTitle = tit
//}


@end

@interface LegacyBackForwardList()
@property (nonatomic, strong) NSMutableArray<LegacyBackForwardListItem*>* writableForwardBackList;
@property (nonatomic) unsigned long currentIndex;
@end


@implementation LegacyBackForwardList

- (void)goBack
{
  if (self.currentIndex > 0)
    self.currentIndex--;
}

- (void)goForward
{
  if (self.currentIndex < self.writableForwardBackList.count - 1)
    self.currentIndex++;
}

- (void)pushItem:(LegacyBackForwardListItem *)item
{
  if (self.writableForwardBackList.count > self.currentIndex + 2) {
    [self.writableForwardBackList removeObjectsInRange:
     NSMakeRange(self.currentIndex + 1, self.writableForwardBackList.count - self.currentIndex + 1)];
  }
  [self.writableForwardBackList addObject:item];
  self.currentIndex = self.writableForwardBackList.count - 1;
}

- (NSMutableArray*)writableForwardBackList
{
  if (!_writableForwardBackList) {
    _writableForwardBackList = [NSMutableArray array];
  }
  return _writableForwardBackList;
}

- (WKBackForwardListItem*)currentItem
{
  return [self itemAtIndex:self.currentIndex];
}

- (WKBackForwardListItem*)backItem
{
  return [self itemAtIndex:self.currentIndex - 1];
}

- (WKBackForwardListItem*)forwardItem
{
  return [self itemAtIndex:self.currentIndex + 1];
}

- (WKBackForwardListItem*)itemAtIndex:(NSInteger)index
{
  LegacyBackForwardListItem* item = (self.writableForwardBackList.count > 0 &&
         index < self.writableForwardBackList.count &&
         index > -1) ?
         self.writableForwardBackList[index] : nil;
  return item;
}

- (NSArray<WKBackForwardListItem*>*)backList
{
  return [NSArray array];
}

-(NSArray<WKBackForwardListItem*>*) forwardList
{
  return [NSArray array];
}

@end

File: Client/LegacyWebViewBackForward.swift
import Foundation

class LegacyBackForwardListItem  {

  var URL: NSURL = NSURL()
  var initialURL: NSURL = NSURL()
  var title:String = ""
}

class LegacyBackForwardList {

  var currentIndex: Int = 0
  var backForwardList: [LegacyBackForwardListItem] = []

  func dump() {
    for i in backList {
      print("\(i.URL.absoluteString)")
    }
  }

  func goBack() {
    if currentIndex > 0 {
      self.currentIndex--
    }
    dump()
  }

  func goForward() {
    if currentIndex < (backForwardList.count - 1) {
      self.currentIndex++
    }
    dump()
  }

  private func isSpecial(_url: NSURL?) -> Bool {
    guard let url = _url else { return false }
    return url.absoluteString.rangeOfString(WebServer.sharedInstance.base) != nil
  }

  func push(_url: NSURL?) {
    guard let url = _url else { return }
    if (url.absoluteString.characters.count < 1 || currentItem?.URL.absoluteString == url.absoluteString) {
      return
    }

    // only one localhost entry allowed, just update the curent one
    if (isSpecial(_url) && isSpecial(currentItem?.URL)) {
      currentItem?.URL = url
      currentItem?.initialURL = url
      return
    }

    let item:LegacyBackForwardListItem = LegacyBackForwardListItem()
    item.URL = url
    item.initialURL = url
    pushItem(item)
  }

  func pushItem(item: LegacyBackForwardListItem) {
    if (backForwardList.count - 1) > currentIndex {
      if let current = currentItem {
        backForwardList = backList
        backForwardList.append(current)
      }
    }
    backForwardList.append(item)
    currentIndex = backForwardList.count - 1
    print("bf count \(backForwardList.count), current index: \(currentIndex) \(item.URL.absoluteString) ##########################")
  }

  var currentItem: LegacyBackForwardListItem? {
    get {
        return itemAtIndex(currentIndex)
    }}

  var backItem: LegacyBackForwardListItem? {
    get {
      return itemAtIndex(currentIndex - 1)
    }}

  var forwardItem: LegacyBackForwardListItem? {
    get {
      return itemAtIndex(currentIndex + 1)
    }}

  func itemAtIndex(index: Int) -> LegacyBackForwardListItem? {
      if (backForwardList.count == 0 ||
          index > backForwardList.count - 1 ||
          index < 0) {
        return nil
      }
      return backForwardList[index]
    }

  var backList: [LegacyBackForwardListItem] {
    get {
      return (currentIndex > 0) ? Array(backForwardList[0..<currentIndex]) : []
    }}

  var forwardList: [LegacyBackForwardListItem] {
    get {
      return (currentIndex + 1 < backForwardList.count) ?
              Array(backForwardList[(currentIndex + 1)..<backForwardList.count]) : []
    }}
}